File: lib/records.js
module.exports = function(lib, xelib, helpers) {
    let { applyEnums, buildFlags, elementContext, Fail, GetArray,
        GetBoolValue, GetStringArray, GetHandle, wcb } = helpers;

    // ENUMERATIONS
    const conflictThis = ['ctUnknown', 'ctIgnored', 'ctNotDefined', 'ctIdenticalToMaster', 'ctOnlyOne', 'ctHiddenByModGroup', 'ctMaster', 'ctConflictBenign', 'ctOverride', 'ctIdenticalToMasterWinsConflict', 'ctConflictWins', 'ctConflictLoses'];
    const conflictAll = ['caUnknown', 'caOnlyOne', 'caNoConflict', 'caConflictBenign', 'caOverride', 'caConflict', 'caConflictCritical'];

    applyEnums(xelib, conflictThis, 'conflictThis');
    applyEnums(xelib, conflictAll, 'conflictAll');

    // FLAGS
    let getRefrsFlags = {
        excludeDeleted: 1,
        excludeDisabled: 2,
        excludeXESP: 4
    };

    // RECORD HANDLING METHODS
    Object.assign(xelib, {
        GetFormID: function(id, native = false, local = false) {
            let _res = Buffer.alloc(4, 0);
            if (!lib.GetFormID(id, _res, native))
                Fail(() => `Failed to get FormID for ${elementContext(id)}`);
            let formID = _res.readUInt32LE();
            return local ? formID & 0xFFFFFF : formID;
        },
        GetHexFormID: function(id, native = false, local = false) {
            return xelib.Hex(xelib.GetFormID(id, native, local), local ? 6 : 8);
        },
        SetFormID: function(id, newFormID, native = false, fixReferences = true) {
            if (!lib.SetFormID(id, newFormID, native, fixReferences))
                Fail(() => `Failed to set FormID on ${elementContext(id)} to ${newFormID}`);
        },
        GetRecord: function(id, formID, searchMasters = true) {
            return GetHandle(function(_res) {
                if (!lib.GetRecord(id, formID, searchMasters, _res))
                    Fail(() => `Failed to get record at: ${elementContext(id)}, ${formID}`);
            });
        },
        GetRecords: function(id, search = '', includeOverrides = false) {
            return GetArray(function(_len) {
                if (!lib.GetRecords(id, wcb(search), includeOverrides, _len))
                    Fail(() => `Failed to get ${search} records from: ${elementContext(id)}`);
            });
        },
        GetREFRs: function(id, search, opts = {}) {
            return GetArray(function(_len) {
                let flags = buildFlags(opts, getRefrsFlags);
                if (!lib.GetREFRs(id, wcb(search), flags, _len))
                    Fail(() => `Failed to get ${search} REFRs from: ${elementContext(id)}`);
            });
        },
        GetOverrides: function(id) {
            return GetArray(function(_len) {
                if (!lib.GetOverrides(id, _len))
                    Fail(() => `Failed to get overrides for: ${elementContext(id)}`);
            });
        },
        GetMasterRecord: function(id) {
            return GetHandle(function(_res) {
                if (!lib.GetMasterRecord(id, _res))
                    Fail(() => `Failed to get master record for: ${elementContext(id)}`);
            });
        },
        GetPreviousOverride: function(id, id2) {
            return GetHandle(function(_res) {
                if (!lib.GetPreviousOverride(id, id2, _res))
                    Fail(() => `Failed to get previous override record for: ${elementContext(id)}, targetting file ${elementContext(id2)}`);
            });
        },
        GetWinningOverride: function(id) {
            return GetHandle(function(_res) {
                if (!lib.GetWinningOverride(id, _res))
                    Fail(() => `Failed to get winning override record for: ${elementContext(id)}`);
            });
        },
        FindNextRecord: function(id, search, byEdid, byName) {
            return GetHandle(function(_res) {
                lib.FindNextRecord(id, wcb(search), byEdid, byName, _res);
            });
        },
        FindPreviousRecord: function(id, search, byEdid, byName) {
            return GetHandle(function(_res) {
                lib.FindPreviousRecord(id, wcb(search), byEdid, byName, _res);
            });
        },
        FindValidReferences: function(id, signature, search, limitTo) {
            return GetStringArray(function(_len) {
                if (!lib.FindValidReferences(id, wcb(signature), wcb(search), limitTo, _len))
                    Fail(() => `Failed to find valid ${signature} references on ${elementContext(id)} searching for: ${search}`);
            });
        },
        GetReferencedBy: function(id) {
            return GetArray(function(_len) {
                if (!lib.GetReferencedBy(id, _len))
                    Fail(() => `Failed to get referenced by for: ${elementContext(id)}`);
            });
        },
        ExchangeReferences: function(id, oldFormID, newFormID) {
            if (!lib.ExchangeReferences(id, oldFormID, newFormID))
                Fail(() => `Failed to exchange references on ${elementContext(id)} from ${oldFormID} to ${newFormID}`)
        },
        IsMaster: function(id) {
            return GetBoolValue(id, "IsMaster");
        },
        IsInjected: function(id) {
            return GetBoolValue(id, "IsInjected");
        },
        IsOverride: function(id) {
            return GetBoolValue(id, "IsOverride");
        },
        IsWinningOverride: function(id) {
            return GetBoolValue(id, "IsWinningOverride");
        },
        GetNodes: function(id) {
            return GetHandle(function(_res) {
                if (!lib.GetNodes(id, _res))
                    Fail(() => `Failed to get nodes for ${elementContext(id)}`);
            });
        },
        GetConflictData: function(id1, id2, asString = false) {
            let _res1 = Buffer.alloc(1, 0),
                _res2 = Buffer.alloc(1, 0);
            if (!lib.GetConflictData(id1, id2, _res1, _res2))
                return [0, 0];
            let n1 = _res1.readUInt8(0),
                n2 = _res2.readUInt8(0);
            return asString ? [conflictAll[n1], conflictThis[n2]] : [n1, n2];
        },
        GetConflictDataEx: function(id1, id2, asString = false) {
            let _res1 = Buffer.alloc(1, 0),
                _res2 = Buffer.alloc(1, 0);
            if (!lib.GetConflictData(id1, id2, _res1, _res2))
                Fail(`GetConflictData failed on ${id1}, ${id2}`);
            let n1 = _res1.readUInt8(0),
                n2 = _res2.readUInt8(0);
            return asString ? [conflictAll[n1], conflictThis[n2]] : [n1, n2];
        },
        GetRecordConflictData: function(id) {
            return xelib.GetConflictData(0, id);
        },
        GetNodeElements: function(id1, id2) {
            return GetArray(function(_len) {
                if (!lib.GetNodeElements(id1, id2, _len))
                    Fail(`GetNodeElements failed on ${elementContext(id1)}, ${elementContext(id2)}`);
            });
        }
    });
};

File: src/cpp/records.cc
#include <nan.h>
#include <windows.h>
#include "records.h"
#include "lib.h"
#include "helpers.h"

using namespace Nan;

NAN_METHOD(GetFormID) {
    if (!TestArguments(info, { atNumber, atBuffer, atBool })) return;
    Cardinal element = GetCardinal(info[0]);
    PCardinal _res = (PCardinal) BufferToPointer(info[1]);
    WordBool native = (WordBool) GetBool(info[2]);
    WordBool success = xelib.functions.GetFormID(element, _res, native);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(SetFormID) {
    if (!TestArguments(info, { atNumber, atNumber, atBool, atBool })) return;
    Cardinal element = GetCardinal(info[0]);
    Cardinal formId = GetCardinal(info[1]);
    WordBool native = (WordBool) GetBool(info[2]);
    WordBool fixRefs = (WordBool) GetBool(info[3]);
    WordBool success = xelib.functions.SetFormID(element, formId, native, fixRefs);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(GetRecord) {
    if (!TestArguments(info, { atNumber, atNumber, atBool, atBuffer })) return;
    Cardinal element = GetCardinal(info[0]);
    Cardinal formId = GetCardinal(info[1]);
    WordBool searchMasters = (WordBool) GetBool(info[2]);
    PCardinal _res = (PCardinal) BufferToPointer(info[3]);
    WordBool success = xelib.functions.GetRecord(element, formId, searchMasters, _res);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(GetRecords) {
    if (!TestArguments(info, { atNumber, atBuffer, atBool, atBuffer })) return;
    Cardinal element = GetCardinal(info[0]);
    PWChar search = (PWChar) BufferToPointer(info[1]);
    WordBool includeOverrides = (WordBool) GetBool(info[2]);
    PInteger len = (PInteger) BufferToPointer(info[3]);
    WordBool success = xelib.functions.GetRecords(element, search, includeOverrides, len);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(GetREFRs) {
    if (!TestArguments(info, { atNumber, atBuffer, atNumber, atBuffer })) return;
    Cardinal element = GetCardinal(info[0]);
    PWChar search = (PWChar) BufferToPointer(info[1]);
    Cardinal flags = GetCardinal(info[2]);
    PInteger len = (PInteger) BufferToPointer(info[3]);
    WordBool success = xelib.functions.GetREFRs(element, search, flags, len);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(GetOverrides) {
    if (!TestArguments(info, { atNumber, atBuffer })) return;
    Cardinal element = GetCardinal(info[0]);
    PInteger len = (PInteger) BufferToPointer(info[1]);
    WordBool success = xelib.functions.GetOverrides(element, len);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(GetReferencedBy) {
    if (!TestArguments(info, { atNumber, atBuffer })) return;
    Cardinal element = GetCardinal(info[0]);
    PInteger len = (PInteger) BufferToPointer(info[1]);
    WordBool success = xelib.functions.GetReferencedBy(element, len);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(GetMasterRecord) {
    if (!TestArguments(info, { atNumber, atBuffer })) return;
    Cardinal element = GetCardinal(info[0]);
    PCardinal _res = (PCardinal) BufferToPointer(info[1]);
    WordBool success = xelib.functions.GetMasterRecord(element, _res);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(GetPreviousOverride) {
    if (!TestArguments(info, { atNumber, atNumber, atBuffer })) return;
    Cardinal e1 = GetCardinal(info[0]);
    Cardinal e2 = GetCardinal(info[1]);
    PCardinal _res = (PCardinal) BufferToPointer(info[2]);
    WordBool success = xelib.functions.GetPreviousOverride(e1, e2, _res);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(GetWinningOverride) {
    if (!TestArguments(info, { atNumber, atBuffer })) return;
    Cardinal element = GetCardinal(info[0]);
    PCardinal _res = (PCardinal) BufferToPointer(info[1]);
    WordBool success = xelib.functions.GetWinningOverride(element, _res);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(GetInjectionTarget) {
    if (!TestArguments(info, { atNumber, atBuffer })) return;
    Cardinal element = GetCardinal(info[0]);
    PCardinal _res = (PCardinal) BufferToPointer(info[1]);
    WordBool success = xelib.functions.GetInjectionTarget(element, _res);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(FindNextRecord) {
    if (!TestArguments(info, { atNumber, atBuffer, atBool, atBool, atBuffer })) return;
    Cardinal element = GetCardinal(info[0]);
    PWChar search = (PWChar) BufferToPointer(info[1]);
    WordBool byEdid = (WordBool) GetBool(info[2]);
    WordBool byName = (WordBool) GetBool(info[3]);
    PCardinal _res = (PCardinal) BufferToPointer(info[4]);
    WordBool success = xelib.functions.FindNextRecord(element, search, byEdid, byName, _res);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(FindPreviousRecord) {
    if (!TestArguments(info, { atNumber, atBuffer, atBool, atBool, atBuffer })) return;
    Cardinal element = GetCardinal(info[0]);
    PWChar search = (PWChar) BufferToPointer(info[1]);
    WordBool byEdid = (WordBool) GetBool(info[2]);
    WordBool byName = (WordBool) GetBool(info[3]);
    PCardinal _res = (PCardinal) BufferToPointer(info[4]);
    WordBool success = xelib.functions.FindPreviousRecord(element, search, byEdid, byName, _res);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(FindValidReferences) {
    if (!TestArguments(info, { atNumber, atBuffer, atBuffer, atNumber, atBuffer })) return;
    Cardinal element = GetCardinal(info[0]);
    PWChar signature = (PWChar) BufferToPointer(info[1]);
    PWChar search = (PWChar) BufferToPointer(info[2]);
    Integer limitTo = GetInt(info[3]);
    PInteger len = (PInteger) BufferToPointer(info[4]);
    WordBool success = xelib.functions.FindValidReferences(element, signature, search, limitTo, len);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(ExchangeReferences) {
    if (!TestArguments(info, { atNumber, atNumber, atNumber })) return;
    Cardinal element = GetCardinal(info[0]);
    Cardinal oldFormId = GetCardinal(info[1]);
    Cardinal newFormId = GetCardinal(info[2]);
    WordBool success = xelib.functions.ExchangeReferences(element, oldFormId, newFormId);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(IsMaster) {
    if (!TestArguments(info, { atNumber, atBuffer })) return;
    Cardinal element = GetCardinal(info[0]);
    PWordBool _res = (PWordBool) BufferToPointer(info[1]);
    WordBool success = xelib.functions.IsMaster(element, _res);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(IsInjected) {
    if (!TestArguments(info, { atNumber, atBuffer })) return;
    Cardinal element = GetCardinal(info[0]);
    PWordBool _res = (PWordBool) BufferToPointer(info[1]);
    WordBool success = xelib.functions.IsInjected(element, _res);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(IsOverride) {
    if (!TestArguments(info, { atNumber, atBuffer })) return;
    Cardinal element = GetCardinal(info[0]);
    PWordBool _res = (PWordBool) BufferToPointer(info[1]);
    WordBool success = xelib.functions.IsOverride(element, _res);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(IsWinningOverride) {
    if (!TestArguments(info, { atNumber, atBuffer })) return;
    Cardinal element = GetCardinal(info[0]);
    PWordBool _res = (PWordBool) BufferToPointer(info[1]);
    WordBool success = xelib.functions.IsWinningOverride(element, _res);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(GetRecordDef) {
    if (!TestArguments(info, { atBuffer, atBuffer })) return;
    PWChar sig = (PWChar) BufferToPointer(info[0]);
    PCardinal _res = (PCardinal) BufferToPointer(info[1]);
    WordBool success = xelib.functions.GetRecordDef(sig, _res);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(GetNodes) {
    if (!TestArguments(info, { atNumber, atBuffer })) return;
    Cardinal element = GetCardinal(info[0]);
    PCardinal _res = (PCardinal) BufferToPointer(info[1]);
    WordBool success = xelib.functions.GetNodes(element, _res);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(GetConflictData) {
    if (!TestArguments(info, { atNumber, atNumber, atBuffer, atBuffer })) return;
    Cardinal e1 = GetCardinal(info[0]);
    Cardinal e2 = GetCardinal(info[1]);
    PByte _res1 = (PByte) BufferToPointer(info[2]);
    PByte _res2 = (PByte) BufferToPointer(info[3]);
    WordBool success = xelib.functions.GetConflictData(e1, e2, _res1, _res2);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(GetNodeElements) {
    if (!TestArguments(info, { atNumber, atNumber, atBuffer })) return;
    Cardinal nodes = GetCardinal(info[0]);
    Cardinal element = GetCardinal(info[1]);
    PInteger _res = (PInteger) BufferToPointer(info[2]);
    WordBool success = xelib.functions.GetNodeElements(nodes, element, _res);
    info.GetReturnValue().Set(Nan::New((bool) success));
}